File: packages/pvm/src/host-functions/accumulate/assign.ts
import { logger } from '@pbnjam/core'
import type { HostFunctionResult, IConfigService } from '@pbnjam/types'
import { ACCUMULATE_FUNCTIONS, RESULT_CODES } from '../../config'
import {
  type AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
} from './base'

/**
 * ASSIGN accumulation host function (Ω_A)
 *
 * Assigns core to service account
 *
 * Gray Paper Specification:
 * - Function ID: 15 (assign)
 * - Gas Cost: 10
 * - Parameters: registers[7-9] = c, o, a
 *   - c: core index
 *   - o: auth queue offset in memory
 *   - a: service account ID to assign
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read auth queue from memory (80 entries * 32 bytes each)
 * 2. Check if core index is valid (< 341)
 * 3. Check if current service is the assigner for this core
 * 4. Check if service account ID is valid
 * 5. Update auth queue and assigner for the core
 */
export class AssignHostFunction extends BaseAccumulateHostFunction {
  readonly functionId = ACCUMULATE_FUNCTIONS.ASSIGN
  readonly name = 'assign'
  readonly gasCost = 10n

  constructor(private readonly configService: IConfigService) {
    super()
  }

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const { registers, ram, implications } = context
    // Extract parameters from registers
    const [coreIndex, authQueueOffset, serviceIdToAssign] = registers.slice(
      7,
      10,
    )

    // Gray Paper constants
    const C_AUTH_QUEUE_SIZE = 80n // Cauthqueuesize
    const C_CORE_COUNT = this.configService.numCores // Ccorecount
    const MAX_SERVICE_ID = 4294967296n // 2^32 = 4294967296

    // Log all input parameters
    logger.info('ASSIGN host function invoked', {
      coreIndex: coreIndex.toString(),
      authQueueOffset: authQueueOffset.toString(),
      serviceIdToAssign: serviceIdToAssign.toString(),
      currentServiceId: implications[0].id.toString(),
      manager: implications[0].state.manager.toString(),
      numCores: C_CORE_COUNT.toString(),
    })
    const [imX] = implications

    // Read auth queue from memory (80 entries * 32 bytes each)
    // Gray Paper pvm_invocations.tex lines 717-720:
    // q = sequence{memory[o+32i:32] for i in N_Cauthqueuesize} when Nrange(o,32*Cauthqueuesize) ⊆ readable(memory), error otherwise
    const authQueueLength = C_AUTH_QUEUE_SIZE * 32n
    const [authQueueData, faultAddress] = ram.readOctets(
      authQueueOffset,
      authQueueLength,
    )
    // Gray Paper line 722: (panic, registers_7, ...) when q = error
    // Note: reference sets OOB on memory faults before PANIC to indicate why it failed
    if (faultAddress || !authQueueData) {
      logger.debug('[ASSIGN] OOB+PANIC: memory read fault', {
        authQueueOffset: authQueueOffset.toString(),
        authQueueLength: authQueueLength.toString(),
        faultAddress: faultAddress?.toString() ?? 'null',
        hasData: (!!authQueueData).toString(),
      })
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }

    if (serviceIdToAssign > MAX_SERVICE_ID) {
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    // Parse auth queue (32 bytes per entry)
    const authQueue: Uint8Array[] = []
    for (let i = 0; i < Number(C_AUTH_QUEUE_SIZE); i++) {
      const entry = authQueueData.slice(i * 32, (i + 1) * 32)
      authQueue.push(entry)
    }

    // Check if core index is valid
    // Gray Paper: c >= Ccorecount
    if (coreIndex >= C_CORE_COUNT) {
      logger.warn('[ASSIGN] CORE error: invalid core index')
      this.setAccumulateError(registers, 'CORE')
      return {
        resultCode: null, // continue execution
      }
    }

    // Get the current implications context

    // Check if current service is the assigner for this core
    // Gray Paper: imX.id !== imX.state.assigners[c]
    if (imX.id !== imX.state.assigners[Number(coreIndex)]) {
      this.setAccumulateError(registers, 'HUH')
      return {
        resultCode: null, // continue execution
      }
    }

    // Check if service account ID is valid
    // Gray Paper: a not in serviceid (assuming positive service IDs)
    if (serviceIdToAssign < 0n) {
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    // Update auth queue and assigner for the core
    // Gray Paper: imX.state.authqueue[c] = q, imX.state.assigners[c] = a
    for (let i = 0; i < authQueue.length; i++) {
      const authQueueEntry = authQueue[i]
      imX.state.authqueue[Number(coreIndex)][i] = authQueueEntry
    }
    imX.state.assigners[Number(coreIndex)] = serviceIdToAssign

    // Set success result
    this.setAccumulateSuccess(registers)

    // Log in the requested format: [host-calls] [serviceId] ASSIGN(coreIndex, serviceIdToAssign) <- OK
    const logServiceId = imX.id
    logger.info(
      `[host-calls] [${logServiceId}] ASSIGN(${coreIndex}, ${serviceIdToAssign}) <- OK`,
    )

    return {
      resultCode: null, // continue execution
    }
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/assign.rs
//! ASSIGN accumulation host function (Ω_A). Gray Paper: function ID 15.
//! r7..r9 = c, o, a (core index, auth queue offset, service ID to assign). 1:1 with AS assign.ts.

use crate::config::FUNC_ASSIGN;
use crate::host_functions::accumulate::base::{
    self, codes, C_AUTH_QUEUE_SIZE, DEFAULT_NUM_CORES, MAX_SERVICE_ID,
};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

const AUTH_ENTRY_SIZE: u32 = 32;

pub struct AssignHostFunction;

impl AssignHostFunction {
    fn parse_auth_queue(data: &[u8]) -> Vec<Vec<u8>> {
        let mut queue = Vec::with_capacity(C_AUTH_QUEUE_SIZE as usize);
        for i in 0..(C_AUTH_QUEUE_SIZE as usize) {
            let start = i * AUTH_ENTRY_SIZE as usize;
            let end = start + AUTH_ENTRY_SIZE as usize;
            queue.push(data[start..end].to_vec());
        }
        queue
    }
}

impl HostFunction for AssignHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_ASSIGN
    }
    fn name(&self) -> &'static str {
        "assign"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let core_index = context.registers[7];
        let auth_queue_offset = context.registers[8];
        let service_id_to_assign = context.registers[9];

        let _log_service_id = context.service_id.unwrap_or(0);
        let num_cores_val = context.num_cores.unwrap_or(DEFAULT_NUM_CORES);
        crate::host_log!(
            "[hostfn] ASSIGN host function invoked coreIndex={} authQueueOffset={} serviceIdToAssign={} currentServiceId={} numCores={}",
            core_index, auth_queue_offset, service_id_to_assign, _log_service_id, num_cores_val
        );

        let auth_queue_length = u64::from(C_AUTH_QUEUE_SIZE) * u64::from(AUTH_ENTRY_SIZE);
        let read_result = context.ram.read_octets(
            auth_queue_offset as u32,
            auth_queue_length as u32,
        );
        if read_result.fault_address != 0 || read_result.data.is_none() {
            base::set_accumulate_error(context.registers, codes::OOB);
            crate::host_log!(
                "[hostfn] assign PANIC: auth_queue read fault (offset={}, fault_address={})",
                auth_queue_offset, read_result.fault_address
            );
            return HostFunctionResult::panic();
        }
        let auth_queue_data = read_result.data.unwrap();

        let auth_queue = Self::parse_auth_queue(&auth_queue_data);

        if core_index >= u64::from(num_cores_val) {
            base::set_accumulate_error(context.registers, codes::CORE);
            crate::host_log!("[hostfn] ASSIGN CORE error: invalid core index");
            crate::host_log!("[host-calls] [{}] ASSIGN({}, {}) <- CORE", _log_service_id, core_index, service_id_to_assign);
            return HostFunctionResult::continue_execution();
        }

        // AS always has implications.regular; single imX.state (accumulation_state).
        let current_service_id = context
            .service_id
            .expect("assign: accumulation context must have service_id");
        let state = context
            .accumulation_state
            .as_deref_mut()
            .expect("assign: accumulation context must have accumulation_state");

        let core_idx = core_index as usize;
        while state.assigners.len() <= core_idx {
            state.assigners.push(0);
        }
        if current_service_id != u64::from(state.assigners[core_idx]) {
            base::set_accumulate_error(context.registers, codes::HUH);
            crate::host_log_error!(
                "[hostfn] assign: HUH (currentServiceId={} is not assigner for core {}, assigner={})",
                current_service_id, core_index, state.assigners[core_idx]
            );
            return HostFunctionResult::continue_execution();
        }

        // WHO (serviceId >= MAX) before "no service" / "no assign_state" HUH so we return WHO when
        // service_id is invalid even if context lacks service_id/assign_state (matches AS outcome).
        if service_id_to_assign >= MAX_SERVICE_ID {
            base::set_accumulate_error(context.registers, codes::WHO);
            crate::host_log_error!(
                "[hostfn] assign: WHO (invalid serviceIdToAssign={}, max={})",
                service_id_to_assign, MAX_SERVICE_ID
            );
            return HostFunctionResult::continue_execution();
        }

        while state.authqueue.len() <= core_idx {
            state.authqueue.push(vec![]);
        }
        state.authqueue[core_idx] = auth_queue;
        state.assigners[core_idx] = service_id_to_assign as u32;

        base::set_accumulate_success(context.registers, codes::OK);
        crate::host_log!("[host-calls] [{}] ASSIGN({}, {}) <- OK", _log_service_id, core_index, service_id_to_assign);
        HostFunctionResult::continue_execution()
    }
}
